ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java

```diff
@@ -918,6 +959,47 @@ private Page<String> executeListAllValidDatasetsStatement(PreparedStatement stat
     }
   }
 
+  /**
+   * Execute listValidDatasetVersionsSql statement.
+   * @param statement the listValidDatasetVersionsSql statement.
+   * @param accountId the id for the parent account.
+   * @param containerId the id of the container.
+   * @param datasetName the name of the dataset.
+   * @param pageToken the page token.
+   * @param versionSchema the dataset version schema.
+   * @return the page of the dataset versions.
+   * @throws SQLException
+   */
+  private Page<String> executeListAllValidDatasetVersionsStatement(PreparedStatement statement, int accountId,
+      int containerId, String datasetName, String pageToken, Dataset.VersionSchema versionSchema) throws SQLException {
+    ResultSet resultSet = null;
+    try {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, datasetName);
+      statement.setInt(4, DatasetVersionState.READY.ordinal());
+      statement.setString(5, pageToken);
+      statement.setString(6, pageToken);
+      statement.setInt(7, mySqlAccountServiceConfig.listDatasetVersionsMaxResult + 1);
+      resultSet = statement.executeQuery();
+      List<String> entries = new ArrayList<>();
+      String nextContinuationToken = null;
+      int resultIndex = 0;
+      while (resultSet.next()) {
+        long versionValue = resultSet.getLong(VERSION);
+        String version = convertVersionValueToVersion(versionValue, versionSchema);
+        if (resultIndex++ == mySqlAccountServiceConfig.listDatasetVersionsMaxResult) {
+          nextContinuationToken = version;
```

Comment:
If I understand correctly, the version we save in the db is "long" for all three types of versionSchema.
For the continue token, I think we don't need convert to customer readable string token. The native "long" type that we save in the database is good enough. Customer doesn't need to know the meaning of the token.
In this way, we don't need "convert to long before query" or "convert to string before sending token to the customer."

And with this PR, I don't see the code that we "convert the string token before query". Am I missing anything? If it's true, the list version with SEMANTIC won't work, right?

Reply:
I forgot that I store version as long type before. So when page token is semantic, the comparison is comparing a string x.x.x with the long value which is wrong. And I also missed the test for that use case, so I added the new test. Now it should work.